serializer.h

```diff
@@ -1757,6 +1765,122 @@ void serialize(Archive & archive, const polymorphic_wrapper<Type> & object)
     registry_instance.serialize(archive, *object);
 }
 
+/**
+ * Support types introduced in c++17
+ */
+#if __cplusplus > 201402L
+/**
+ * The base function of the variant serialization helper function.
+ */
+template <typename Archive, typename Variant,
+    typename = typename Archive::loading
+>
+void serialize(Archive &, const Variant &, std::size_t)
+{
+}
+
+/**
+ * Serialize std::variant, operates on loading(input) archives.
+ * Enumerate on all the variant types, until the active type is
+ * reached, and serialize it.
+ */
+template <typename Archive, typename VariantItem, typename... VariantItems,
+    typename Variant, typename = typename Archive::loading
+>
+void serialize(Archive & archive, Variant & variant, std::size_t index)
+{
+    if (0 == index) {
+        VariantItem item;
+        archive(item);
+        variant = item;
+        return;
+    }
+
+    serialize<Archive, VariantItems...>(archive, variant, index - 1);
+}
+
+/**
+ * Serialize std::variant, operates on loading (input) archives.
+*/
+template <typename Archive, typename... VariantItems,
+    typename = typename Archive::loading
+>
+void serialize(Archive & archive, std::variant<VariantItems...> & variant)
+{
+    // Serialize the active variant index
+    decltype(variant.index()) active_index;
+    archive(active_index);
+    if (active_index >= sizeof...(VariantItems)) {
+        throw out_of_range("Variant type mismatch");
+    }
+
+    // Serialize the active variant type, if one exists.
+    variant = {};
+    if (std::variant_npos != active_index) {
+        // A helper function for the variant to trigger the archive()
+        // on the active variant index.
+
+        serialize<Archive, VariantItems...>(archive, variant, active_index);
+    }
+}
+
+/**
+ * Serialize std::variant, operates on saving (output) archives.
+ */
+template <typename Archive, typename... VariantItems,
+    typename = typename Archive::saving
+>
+void serialize(Archive & archive, const std::variant<VariantItems...> & variant)
+{
+    auto variant_index = variant.index();
+
+    // Serialize the variant index, to tell which variant type was active.
+    archive(variant_index);

```

Comment:
Can save the index in a cross platform manner? I'd choose uint32_t as the type, similar to size_type that was chosen to be 32 bit.

Reply:
Changed it to std::uint32_t.
std::size_t is implictly casted to std::uint32_t, even though data might be lost on 64bit platforms?

Reply:
Yes I chose 32 bit assuming nobody is going to serialize something with more than 32 bit size (imagine that), and it significantly reduces the price of writing 8 byte size every time.

I think the same thing applies in the variant index (variant of 64 bit types, even 32, seems impossible).
I'll probably change to 8/16 bit if it turns out faster than 32.

Reply:
@eyalz800  The problem is std::variant_npos is -1(thats an implementation detail), on 64bit you won't be able to cast it to std::uint32_t, this is why std::uint64_t seems more suitable.

Reply:
We can test npos and write our max integer value for example, when we read that max integer we don't initialize the variant.

Reply:
In any case I'll do some checks in the weekend and change it to whatever I see best.